af/fastpath/fastpath/normalize.py

```diff
@@ -0,0 +1,811 @@
+    value_to_hash += header.get("software_version", "0.0.0").encode("utf-8")
+    probe_city = header.get("probe_city", "None")
+    # probe_city = probe_city.encode("utf-8")  # u'Reykjav\xedk' in bucket 2014-02-20
+    # probe_city = (
+    #   probe_city.encode("utf-8")
+    #   if isinstance(probe_city, unicode)
+    #   else str(probe_city)
+    # )  # u'Reykjav\xedk' in bucket 2014-02-20
+    # probe_city = probe_city.decode() #encode("utf-8")
+    value_to_hash += probe_city.encode("utf-8")
+    report_id += "".join(
+        string.ascii_letters[b % len(string.ascii_letters)]
+        for b in hashlib.sha512(value_to_hash).digest()
+    )[:50]
+    return report_id
+
+
+def iter_yaml_lz4_reports(fn):
+
+    fd = lz4frame.open(fn)
+    report_gen = stream_yaml_blobs(fd)
+
+    off, header = next(report_gen)
+    headsha = hashlib.sha1(header)
+    # XXX: bad header kills whole bucket
+    header = yaml.load(header, Loader=CLoader)
+    if not header.get("report_id"):
+        header["report_id"] = generate_report_id(header)
+
+    for off, entry in report_gen:
+        entry_len = len(entry)
+        esha = headsha.copy()
+        esha.update(entry)
+        esha = esha.digest()
+        try:
+            entry = yaml.load(entry, Loader=CLoader)
+            if not entry:  # e.g. '---\nnull\n...\n'
+                continue
+            if "test_start_time" in entry and "test_start_time" in header:
+                header.pop("test_start_time")
+            entry.update(header)
+            yield off, entry_len, esha, entry, None
+        except Exception as exc:
+            yield off, entry_len, esha, None, exc
+
+    fd.close()
+
+
+def iter_yaml_msmt_normalized(data):
+    """Yields normalized measurements from a YAML BytesIO
+    """
+    report_gen = stream_yaml_blobs(data)
+
+    off, header = next(report_gen)
+    headsha = hashlib.sha1(header)
+    # XXX: bad header kills whole bucket
+    header = yaml.load(header, Loader=CLoader)
+    if not header.get("report_id"):
+        header["report_id"] = generate_report_id(header)
+
+    for off, entry in report_gen:
+        entry_len = len(entry)
+        esha = headsha.copy()
+        esha.update(entry)
+        esha = esha.digest()
+        try:
+            entry = yaml.load(entry, Loader=CLoader)
+            if not entry:  # e.g. '---\nnull\n...\n'
+                continue
+            if "test_start_time" in entry and "test_start_time" in header:
+                header.pop("test_start_time")
+            entry.update(header)
+            yield normalize_entry(entry, "2018-05-07", "??", esha)
+        except Exception as exc:
+            yield normalize_entry(entry, "2018-05-07", "??", esha)
+
+
+def iter_yaml_lz4_reports_normalized(fn):
+    """Yields normalized measurements from a yaml.lz4 file
+    """
+    for r in iter_yaml_lz4_reports(fn):
+        off, entry_len, esha, entry, exc = r
+        yield normalize_entry(entry, "2018-05-07", "??", esha)
```

Comment:
Perhaps these magic numbers `2018-05-07` and `??` should be defined as a constant variable and set to a value which is more clearly something hardcoded true (for example `2010-01-02` before there were any OONI measurements).

Reply:
See comment above.